Don't internalize __llvm_profile_counter_bias

abrachet opened the pull request:
Currently, LLVM profiling runtime counter relocation cannot be used by rust during LTO because symbols are being internalized before all symbol information is known.

This mode makes LLVM emit a __llvm_profile_counter_bias symbol which is referenced by the profiling initialization, which itself is pulled in by the rust driver here [1].

It is enabled with -Cllvm-args=-runtime-counter-relocation for platforms which are opt-in to this mode like Linux. On these platforms there will be no link error, rather just surprising behavior for a user which request runtime counter relocation. The profiling runtime will not see that symbol go on as if it were never there. On Fuchsia, the profiling runtime must have this symbol which will cause a hard link error.

As an aside, I don't have enough context as to why rust's LTO model is how it is. AFAICT, the internalize pass is only safe to run at link time when all symbol information is actually known, this being an example as to why. I think special casing this symbol as a known one that LLVM can emit which should not have it's visbility de-escalated should be fine given how seldom this pattern of defining an undefined symbol to get initilization code pulled in is. From a quick grep, __llvm_profile_runtime is the only symbol that rustc does this for.

[1] https://github.com/rust-lang/rust/blob/0265a3e93bf1b89d97cae113ed214954d5c35e22/compiler/rustc_codegen_ssa/src/back/linker.rs#L598

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @cjgillot (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

bjorn3 commented:
I think it should be added to https://github.com/rust-lang/rust/blob/bb93450ec4af83c20e9ba6c8e575aca55423001a/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L216 instead.

abrachet commented:
> You can squash commits to remove now unneeded changes.

Done

cjgillot commented:
r? @bjorn3

bjorn3 commented:
Could you add a test for this?

abrachet commented:
> Could you add a test for this?

Sure, let me know if that test looks ok.

bjorn3 commented on src/test/codegen/pgo-counter-bias.rs:
```diff
@@ -0,0 +1,9 @@
+// Test that __llvm_profile_counter_bias does get internalized by lto.
```
*does not?

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 2f7b4d91e21863cd7e9ffc24d18299db432639c5 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 2f7b4d91e21863cd7e9ffc24d18299db432639c5 -->
<!-- homu: {"type":"Approved","sha":"2f7b4d91e21863cd7e9ffc24d18299db432639c5","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2f7b4d91e21863cd7e9ffc24d18299db432639c5 with merge a16abce7a31818d541760097059538aa6c5aa1e4...
<!-- homu: {"type":"BuildStarted","head_sha":"2f7b4d91e21863cd7e9ffc24d18299db432639c5","merge_sha":"a16abce7a31818d541760097059538aa6c5aa1e4"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3259537747/jobs/5352633000)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3259537747/jobs/5352633000","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3259537747/jobs/5352439057) [(plain)](https://github.com/rust-lang-ci/rust/commit/a16abce7a31818d541760097059538aa6c5aa1e4/checks/8916567931/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [mir-opt] src/test\mir-opt\coverage_graphviz.rs stdout ----

error: compilation failed!
status: exit code: 1
command: PATH="D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\bin;D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0-bootstrap-tools\x86_64-pc-windows-gnu\release\deps;D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage0\bin;D:\a\rust\rust\ninja;D:\a\rust\rust\mingw64\bin;C:\hostedtoolcache\windows\Python\3.10.7\x64\Scripts;C:\hostedtoolcache\windows\Python\3.10.7\x64;C:\msys64\usr\bin;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\cf-cli;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.9.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\tools\ghc-9.4.2\bin;C:\Program Files\dotnet;C:\mysql\bin;C:\Program Files\R\R-4.2.1\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.17.13\x64\bin;C:\hostedtoolcache\windows\Python\3.7.9\x64\Scripts;C:\hostedtoolcache\windows\Python\3.7.9\x64;C:\hostedtoolcache\windows\Ruby\2.5.9\x64\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.345-1\x64\bin;C:\Program Files\ImageMagick-7.1.0-Q16-HDRI;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\ProgramData\Chocolatey\bin;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\dotnet;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn;C:\Program Files\OpenSSL\bin;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\TortoiseSVN\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.6\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\SeleniumWebDrivers\ChromeDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files (x86)\Google\Cloud SDK\google-cloud-sdk\bin;C:\Program Files (x86)\Microsoft BizTalk Server;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\bin\\rustc.exe" "D:\\a\\rust\\rust\\src/test\\mir-opt\\coverage_graphviz.rs" "-Zthreads=1" "--target=x86_64-pc-windows-gnu" "-O" "-Copt-level=1" "-Zdump-mir=all" "-Zvalidate-mir" "-Zdump-mir-exclude-pass-number" "-Zmir-pretty-relative-line-numbers=yes" "-Zmir-opt-level=4" "-Zdump-mir-dir=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz" "-C" "prefer-dynamic" "--out-dir" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz" "-Crpath" "-Cdebuginfo=0" "-Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\native\\rust-test-helpers" "-C" "instrument-coverage" "-Z" "dump-mir-graphviz" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\auxiliary"
stdout: none
--- stderr -------------------------------
error: linking with `x86_64-w64-mingw32-gcc` failed: exit code: 1
  |
  = note: "x86_64-w64-mingw32-gcc" "-fno-use-linker-plugin" "-Wl,--dynamicbase" "-Wl,--disable-auto-image-base" "-m64" "-Wl,--high-entropy-va" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustc6larzQ\\symbols.o" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\coverage_graphviz.coverage_graphviz.0c8b2060-cgu.0.rcgu.o" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\coverage_graphviz.coverage_graphviz.0c8b2060-cgu.1.rcgu.o" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\coverage_graphviz.coverage_graphviz.0c8b2060-cgu.2.rcgu.o" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\native\\rust-test-helpers" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\auxiliary" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libprofiler_builtins-f75151427eb85218.rlib" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bdynamic" "-lstd-3ab0047ea5e7d241" "-Wl,-Bstatic" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-829871d8657f04c6.rlib" "-Wl,-Bdynamic" "-lkernel32" "-ladvapi32" "-luserenv" "-lkernel32" "-lws2_32" "-lbcrypt" "-lgcc_s" "-lmsvcrt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "-Wl,--nxcompat" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-o" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\test\\mir-opt\\coverage_graphviz\\coverage_graphviz.exe" "-Wl,--gc-sections" "-no-pie" "-nodefaultlibs" "-u" "__llvm_profile_runtime" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage2\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\lib\rustlib\x86_64-pc-windows-gnu\lib\libprofiler_builtins-f75151427eb85218.rlib(InstrProfilingFile.o):InstrProfilingFile.c:(.text$mmapForContinuousMode.isra.1+0xd3): undefined reference to `__llvm_profile_counter_bias'
          D:\a\rust\rust\build\x86_64-pc-windows-gnu\stage2\lib\rustlib\x86_64-pc-windows-gnu\lib\libprofiler_builtins-f75151427eb85218.rlib(InstrProfilingFile.o):InstrProfilingFile.c:(.text$initializeProfileForContinuousMode+0x2b): undefined reference to `__llvm_profile_counter_bias'
          collect2.exe: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error
------------------------------------------


---
test result: FAILED. 183 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 18.96s

Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-pc-windows-gnu target=x86_64-pc-windows-gnu
Build completed unsuccessfully in 0:36:28
make: *** [Makefile:83: ci-mingw-subset-1] Error 1
```

</details>

abrachet commented:
It looks like this creates an undefined reference to these symbols (in /tmp/.../symbols.o), which is fine on other platforms but I guess doesn't work so well with how weak symbols are implemented in Windows, see https://github.com/llvm/llvm-project/blob/6db71b8f1418170324b49d20f1f7b3f7c5086066/compiler-rt/lib/profile/InstrProfilingFile.c#L168-L178.

As far as I can tell the fix is to stop indiscriminately adding this to the exported symbols list and only do so when `__llvm_profile_counter_bias` is emitted, which seems untenable given llvm decides to do this based on target (which currently is only one target but could likely grow) or backend flag. Alternatively we could go back to putting this in `prepare_lto`, though here is admittedly much better because this is where other similar symbols are handled. WDYT?

bjorn3 commented:
I'm not quite sure. Maybe @nikic has an idea?

Manishearth commented:
@bors r- 

somehow got readded

abrachet commented:
@bjorn3 what do you think about moving back to adding `__llvm_profile_counter_bias` to `symbols_below_threshold` here [1]? Seems like that is the cleanest way given the problem described above on Windows. WDYT?


[1] https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/back/lto.rs#L135

nikic commented:
Sounds reasonable to me.

abrachet committed:
Don't internalize __llvm_profile_counter_bias

Currently, LLVM profiling runtime counter relocation cannot be
used by rust during LTO because symbols are being internalized
before all symbol information is known.

This mode makes LLVM emit a __llvm_profile_counter_bias symbol
which is referenced by the profiling initialization, which itself
is pulled in by the rust driver here [1].

It is enabled with -Cllvm-args=-runtime-counter-relocation for
platforms which are opt-in to this mode like Linux. On these
platforms there will be no link error, rather just surprising
behavior for a user which request runtime counter relocation.
The profiling runtime will not see that symbol go on as if it
were never there. On Fuchsia, the profiling runtime must have
this symbol which will cause a hard link error.

As an aside, I don't have enough context as to why rust's LTO
model is how it is. AFAICT, the internalize pass is only safe
to run at link time when all symbol information is actually
known, this being an example as to why. I think special casing
this symbol as a known one that LLVM can emit which should not
have it's visbility de-escalated should be fine given how
seldom this pattern of defining an undefined symbol to get
initilization code pulled in is. From a quick grep,
__llvm_profile_runtime is the only symbol that rustc does this
for.

[1] https://github.com/rust-lang/rust/blob/0265a3e93bf1b89d97cae113ed214954d5c35e22/compiler/rustc_codegen_ssa/src/back/linker.rs#L598

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 615d3956cc098d359da8f98ff768efbbc3ddd7d3 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 615d3956cc098d359da8f98ff768efbbc3ddd7d3 -->
<!-- homu: {"type":"Approved","sha":"615d3956cc098d359da8f98ff768efbbc3ddd7d3","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 615d3956cc098d359da8f98ff768efbbc3ddd7d3 with merge 906637d90c9e0f53bab59082ec955a62fb51ffba...
<!-- homu: {"type":"BuildStarted","head_sha":"615d3956cc098d359da8f98ff768efbbc3ddd7d3","merge_sha":"906637d90c9e0f53bab59082ec955a62fb51ffba"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3637350027/jobs/6138408884)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3637350027/jobs/6138408884","builder_name":"checks-actions"} -->

nikic commented:
@bors retry network error

bors commented:
:hourglass: Testing commit 615d3956cc098d359da8f98ff768efbbc3ddd7d3 with merge 1137832ff947136b3a2d73bb9760793f7e660967...
<!-- homu: {"type":"BuildStarted","head_sha":"615d3956cc098d359da8f98ff768efbbc3ddd7d3","merge_sha":"1137832ff947136b3a2d73bb9760793f7e660967"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3637718547/jobs/6139634959)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3637718547/jobs/6139634959","builder_name":"checks-actions"} -->

abrachet commented:
Looks like `-runtime-counter-relocation` is ignored for Mach-O targets [1]. I've added an `// ignore-macos` line in the test which should hopefully fix the test failure on the bots. Sorry about that

[1] https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Instrumentation/InstrProfiling.cpp#L443

bjorn3 commented:
@bors r+

bors commented:
:hourglass: Testing commit 5d88d3605339d61c81d16b29a1c01edf771a8fe6 with merge d31d00eef1d034f6e478f0088c74516ae3a98f44...
<!-- homu: {"type":"BuildStarted","head_sha":"5d88d3605339d61c81d16b29a1c01edf771a8fe6","merge_sha":"d31d00eef1d034f6e478f0088c74516ae3a98f44"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3643606003/jobs/6152236977)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3643606003/jobs/6152236977","builder_name":"checks-actions"} -->

ehuss commented:
@bors retry

Failed to update toolstate 
(has been fixed)

bjorn3 commented:
@bors retry Failed to update toolstate

bjorn3 commented:
@bors retry

ehuss commented:
This PR is already in the queue. The rust-log-analyzer comment was just delayed by about 12 hours.

bors commented:
:hourglass: Testing commit 5d88d3605339d61c81d16b29a1c01edf771a8fe6 with merge d137783642b0b98eda2796dc66bffc2b089a8327...
<!-- homu: {"type":"BuildStarted","head_sha":"5d88d3605339d61c81d16b29a1c01edf771a8fe6","merge_sha":"d137783642b0b98eda2796dc66bffc2b089a8327"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3669456292/jobs/6204355377)
Approved by: bjorn3
Pushing d137783642b0b98eda2796dc66bffc2b089a8327 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3669456292/jobs/6204355377"},"merge_sha":"d137783642b0b98eda2796dc66bffc2b089a8327"} -->

rust-timer commented:
Finished benchmarking commit (d137783642b0b98eda2796dc66bffc2b089a8327): [comparison URL](https://perf.rust-lang.org/compare.html?start=4de4d60779ba9c1f79d851e8ffceb038e9042610&end=d137783642b0b98eda2796dc66bffc2b089a8327&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=4de4d60779ba9c1f79d851e8ffceb038e9042610&end=d137783642b0b98eda2796dc66bffc2b089a8327&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4de4d60779ba9c1f79d851e8ffceb038e9042610&end=d137783642b0b98eda2796dc66bffc2b089a8327&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.0%, 2.7%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.7%, -2.3%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4de4d60779ba9c1f79d851e8ffceb038e9042610&end=d137783642b0b98eda2796dc66bffc2b089a8327&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.5% | [2.5%, 2.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>
